nrf_rpc/doc/architecture.rst

```diff
@@ -0,0 +1,137 @@
+.. _nrf_rpc_architecture:
+
+Architecture
+############
+
+The following picture gives an overview of the nRF RPC architecture:
+
+.. image:: img/layers.svg
+   :alt: nRF RPC Architecture Layers
+   :align: center
+
+
+Types of communication
+======================
+
+You can issue the calls in one of two ways: using a command/response or an event.
+
+A **command and response** is intended to be used for synchronous function calls.
+The caller sends the command to the other side and waits for a response.
+The response is sent after the remote function returns on the remote core.
+The following image shows an example of the command and response flow.
+
+.. image:: img/cmd_flow.svg
+   :alt: nRF RPC Command/Response flow
+   :align: center
+
+**Events** are intended to be used for asynchronous function calls.
+The caller sends an event to the other side and returns immediately if there is a free thread in the thread pool (see the :ref:`nrf_rpc_architecture_threads` section).
+Otherwise, it waits for an available thread.
+It is not possible to return anything from the remote side after receiving an event, but it is possible to send an event in the opposite direction.
+The following image shows an example of the event flow.
+
+.. image:: img/evt_flow.svg
+   :alt: nRF RPC Event flow
+   :align: center
+
+.. _nrf_rpc_architecture_threads:
+
+Threads
+=======
+
+When a remote procedure call is received, it must be executed in context of a thread.
+For this reason, each side of the nRF RPC contains a thread pool.
+It is OS-dependent and implemented in the OS abstraction layer.
+The number of threads in a thread pool is configurable.
+
+When a caller wants to call a remote procedure, the nRF RPC checks if there is a thread available in a pool.
+If there is not, then it waits until one becomes available, blocking the caller.
+After that, nRF RPC sends a packet and it is executed by an available thread from a thread pool.
+In the case of a command being received, a response is sent directly to a waiting thread, so no new thread is allocated for a response.
+
+The following image presents sample command/response diagrams.
+
+.. image:: img/cmd_simple.svg
+   :alt: nRF RPC Simple Command flow
+   :align: center
+
+A thread waiting for a response may be reused to receive a new incoming command from the remote thread that the local thread is waiting for, e.g. when a callback is called synchronously.
```

Comment:
Is it possible to get some code example for this?
I am struggling to understand how a thread can wait and process a callback as the same time.

Reply:
Is it an option that another RPC call is executed from that synchronous callback?

Reply:
> Is it possible to get some code example for this?

There is no example of this at the moment. Do you think that it will be beneficial for the user to show this feature in the "entropy" sample (https://github.com/nrfconnect/sdk-nrf/pull/2179)?

> I am struggling to understand how a thread can wait and process a callback as the same time. Is it supposed to be based on stream from the remote? E.g. the remote says to call a callback as part of its result.

You can look at `nrf_rpc/doc/img/cmd_recursive.svg`. Simple description how it works:
1. Local thread sends a command 1 and waits for incoming packets i.e. a response packet or another command packet.
2. Remote thread receives the command and executes it.
    * During its execution command 2 is send from the remote thread.
    * nRF RPC knows that we are executing command 1 that was requested by the local thread, so it sends it directly to the local thread.
3. Local thread received packet with command 2, so it executes it in context of the same thread and sends a response to remote thread.
4. Remote thread completes execution of the command 1 and sends a response packet to the local thread.
5. Local thread receives the response, so it returns back from the RPC to user code.

> Is it an option that another RPC call is executed from that synchronous callback?

If you asking if a thread waiting for a response is able to execute other commands than described above then the answer is no.

Reply:
I find the examples for RPC extremely useful. Therefore, if it is possible, I think it is worth having it.

Thank you for details on how recursive case works. I have a follow-up question: is there a limit on depth in recursive scenario?
E.g., https://github.com/nrfconnect/sdk-zephyr/blob/master/include/bluetooth/bluetooth.h#L1053
This is Zephyr host public API and a candidate to be an RPC function. As you can see this function is called from a callback (RPC call #1), then the application calls this function (RPC call #2) which can also call passed callback on remote (call #3). Or how will this work?

Reply:
Recursive depth is limited by the stack size. Scenario with the multiple, callbacks that you shown above will work the same as in the single core case (at least from a user perspective).

Small comment regarding `bt_data_parse` function. I have different plans for it. When you look at the function body you will see that it is very simple and independent. Overhead required to encode and decode data for this function will be greater than the function itself, so I thinking that this function should be implemented in the application core without any RPC.

Reply:
I added recursive commands into the sample:

https://github.com/nrfconnect/sdk-nrf/pull/2179/commits/fa0110d1ea5ed0887f7be71764c2f3203cac3c6d